plugins/modules/source_control/gitlab/gitlab_project_variable.py

```diff
@@ -162,106 +243,111 @@ def list_all_project_variables(self):
+
+    # preprocessing:filter out and enrich before compare
+    for item in existing_variables:
+        item.pop('project_id')
+
+    for item in requested_variables:
+        item['key'] = item.pop('name')
+        item['value'] = str(item.get('value'))
+        if item.get('protected') is None:
+            item['protected'] = False
+            module.deprecate("The 'protected' attribute will be set to 'True' if not defined.",
+                             version='5.0.0', collection_name='community.general')
+        if item.get('masked') is None:
+            item['masked'] = False
+        if item.get('environment_scope') is None:
+            item['environment_scope'] = '*'
+        if item.get('variable_type') is None:
+            item['variable_type'] = 'env_var'
+
+    if state == 'present':
+        add_or_update = [x for x in requested_variables if x not in existing_variables]
+        for item in add_or_update:
+            try:
+                if this_gitlab.create_variable(item):
+                    return_value['added'].append(item)
+
+            except Exception:
+                if this_gitlab.update_variable(item):
+                    return_value['updated'].append(item)
+
+        if purge:
+            # refetch and filter
+            gitlab_keys = this_gitlab.list_all_project_variables()
+            existing_variables = [x.attributes for x in gitlab_keys]
+            for item in existing_variables:
+                item.pop('project_id')
+
+            remove = [x for x in existing_variables if x not in requested_variables]
+            for item in remove:
+                if this_gitlab.delete_variable(item):
+                    return_value['removed'].append(item)
+
+    elif state == 'absent':
+        # value does not matter on removing variables.
+        # key and environment scope are sufficient
         for item in existing_variables:
-            this_gitlab.delete_variable(item)
-            change = True
-            return_value['removed'].append(item)
-    else:
-        return_value['untouched'].extend(existing_variables)
+            item.pop('value')
+            item.pop('variable_type')
+        for item in requested_variables:
+            item.pop('value')
+            item.pop('variable_type')
+
+        if not purge:
+            remove_requested = [x for x in requested_variables if x in existing_variables]
+            for item in remove_requested:
+                if this_gitlab.delete_variable(item):
+                    return_value['removed'].append(item)
 
-    return change, return_value
+        else:
+            for item in existing_variables:
+                if this_gitlab.delete_variable(item):
+                    return_value['removed'].append(item)
+
+    if len(return_value['added'] + return_value['removed'] + return_value['updated']) > 0:
+        change = True
+
+    gitlab_keys = this_gitlab.list_all_project_variables()
+    after = [x.attributes for x in gitlab_keys]
```

Comment:
I don't think this works in check mode.

Reply:
Do mean because `before` and `after` will have the same values in check mode?

That do you suggest instead? No return of diff in check mode?

Reply:
> Do mean because `before` and `after` will have the same values in check mode?

Yes. In the old version this was correctly returned also in check mode (IIRC).

> That do you suggest instead? No return of diff in check mode?

I would probably compute the value the same way as it was computed in the previous version.

Reply:
> In the old version ...

The old version was spaghetti code from hell :D But I see what I can do.

Reply:
The behaviour is worth discussing 

Say there are two variables: same name, but different scrope.  
Now we request a third variable in purge mode: also the same name, and also a different env scope.

The current implementation counts one delete and one update

```
ok: [localhost] => {
    "gitlab_project_variable_state": {
        "changed": true,
        "failed": false,
        "project_variable": {
            "added": [],
            "removed": [
                "SECRET_ACCESS_KEY"
            ],
            "untouched": [],
            "updated": [
                "SECRET_ACCESS_KEY"
            ]
        }
    }
}
```

Technical that's impossible. We can change the value, but nothing else.  
The new implementation I got now, counts two removals and one add - what is technical correct. The current implementation is a bug.


But I think we want to keep the bug behaviour of an update state for the users, right? :D

Reply:
It's not possible to implement the `updated` count like it was before. Because the current implementation does not respect different environment scopes using the same variable name. Therefore the `update` might happen in the future. But in the example above, the current implementation is a bug.

Reply:
The changed behavior sounds good to me. It should be documented in the changelog though.